enterprise/api/v1/views.py

```diff
@@ -577,17 +574,71 @@ def license_revoke(self, request, *args, **kwargs):
             course_id = str(course_overview.get('id'))
             enterprise_enrollment = enrollments_by_course_id.get(course_id)
             try:
-                revocation_status = self._revoke_enrollment(
+                revocation_status = self._terminate_enrollment(
                     enrollment_api_client, enterprise_enrollment, course_overview
                 )
                 revocation_results[course_id] = {'success': True, 'message': revocation_status}
+                if revocation_status != self.EnrollmentTerminationStatus.COURSE_COMPLETED:
+                    enterprise_enrollment.license.revoke()
             except EnrollmentModificationException as exc:
                 revocation_results[course_id] = {'success': False, 'message': str(exc)}
                 any_failures = True
+                enterprise_enrollment.license.revoke()
 
         status_code = status.HTTP_200_OK if not any_failures else status.HTTP_422_UNPROCESSABLE_ENTITY
         return Response(revocation_results, status=status_code)
 
+    @action(methods=['post'], detail=False)
+    @permission_required('enterprise.can_enroll_learners')
+    def licensed_enrollments_expiration(self, request):
+        """
+        Changes the mode for licensed enterprise course enrollments to the "audit" course mode,
+        or unenroll the user if no audit mode exists for each expired license uuid
+        """
+        expired_license_uuids = get_request_value(request, self.REQ_EXP_LICENSE_UUIDS_PARAM, '')
+
+        if not expired_license_uuids:
+            return Response(
+                'Parameter {} must be provided'.format(self.REQ_EXP_LICENSE_UUIDS_PARAM),
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        licensed_enrollments = models.LicensedEnterpriseCourseEnrollment.objects.filter(
+            license_uuid__in=expired_license_uuids
+        ).select_related('enterprise_course_enrollment')
+
+        enrollment_api_client = EnrollmentApiClient()
+        course_overviews = get_course_overviews(
+            list(licensed_enrollments.values_list('enterprise_course_enrollment__course_id', flat=True))
+        )
+        indexed_overviews = {overview.get('id'): overview for overview in course_overviews}
+
+        any_failures = False
+        for licensed_enrollment in licensed_enrollments:
+            enterprise_course_enrollment = licensed_enrollment.enterprise_course_enrollment
+            course_id = enterprise_course_enrollment.course_id
+            course_overview = indexed_overviews.get(course_id)
+            try:
+                termination_status = self._terminate_enrollment(
```

Comment:
I think there's a limit to the number of API calls we can make at a time. If a subscription is going to have a max 5k learners and each learner could have multiple enrollments then it might make sense to set a delay for every few hundred enrollments

Reply:
I would just make this endpoint return an error if `expired_license_uuids` is longer than some constant, 100 is probably fine.

Reply:
I'm also interested to see how badly `_terminate_enrollment()` slows us down when we try to expire 100s of enrollments! (I'm not suggesting you change anything, here, just expressing genuine curiosity)